ec_api/apps/api_proxy/tests/response_builder.py

```diff
@@ -0,0 +1,256 @@
+                "polling_station_known": False,
+                "custom_finder": None,
+                "report_problem_url": None,
+                "station": None,
+            },
+            "notifications": [],
+            "ballots": BallotsResponse(),
+        }
+
+    def with_future_dates(self, count=1, date_list: list = None):
+        if date_list:
+            assert isinstance(date_list, list)
+        assert not self._dates, "Dates already set"
+        assert not self.options.get("split_council"), (
+            "No dates shown for split " "postcodes"
+        )
+        from_date = datetime.date.today()
+        if not date_list:
+            date_list = []
+            for i in range(count):
+                random_days_in_future = randrange(5, 30)
+                election_date = from_date + datetime.timedelta(
+                    days=random_days_in_future
+                )
+                date_list.append(election_date)
+                from_date = election_date
+
+        dates = []
+        for date in date_list:
+            dates.append(self.get_date_object(date))
+
+        self._dates = dates
+        return self
+
+    def with_ballot_on_date(self, date, ballot):
+        if not self.get_dates:
+            self.with_future_dates(date_list=[date])
+        assert date in [d["date"] for d in self.get_dates]
+        for date_obj in self.get_dates:
+            if date_obj["date"] == date:
+                date_obj["ballots"].with_ballot(ballot)
+        return self
+
+
+class PostcodeResponse(DatesMixin, CouncilContactDetailsMixin, BaseResponse):
+    def get_response(self):
+        return {
+            "address_picker": self.get_address_picker,
+            "addresses": self.get_addresses,
+            "dates": self.get_dates,
+            "postcode_location": self.get_postcode_location,
+            "electoral_services": self.get_electoral_services,
+            "registration": self.get_registration,
+        }
+
+    @property
+    def get_addresses(self):
+        return []
+
+    @property
+    def get_address_picker(self):
+        return getattr(self, "address_picker", False)
+
+    def with_addresses(self, number=10):
+        self.address_picker = True
+        # del council contacts
+        return self
+
+    @property
+    def get_postcode_location(self):
+        return {
+            "type": "Feature",
+            "properties": None,
+            "geometry": {
+                "type": "Point",
+                "coordinates": [-0.13447605, 51.489488200000004],
+            },
+        }
+
+    def with_split_over_councils_postcode(self):
+        """
+        This postcode is split over two or more councils
+        """
+        self.options["split_council"] = True
+        self.with_addresses(5)
+        return self
+
+    def build_response(self):
+        self._response = self.get_response()
+        self._response["addresses"] = self.get_addresses
```

Comment:
Looks like these key/values already get set within `get_response`, are they needed here as well?

Reply:
Actually `get_response` isn't named well - the idea is that `get_response` is the _initial_ empty response that's used to populate `_response` when the class is instantiated.

When we call a `with_*` method, it will add data to the response but also it might change some other data. I think this means that it's best to call `get_*` methods right before returning a response.

Maybe this isn't right though, and we can just rely on building `_response` over time...I'll have a think